nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/PutKinesis.java

```diff
@@ -0,0 +1,174 @@
+           AWS_KINESIS_ERROR_MESSAGE);
+
+        final AmazonKinesisClient client = getClient();
+
+        try {
+            List<PutRecordsRequestEntry> records = new ArrayList<>();
+
+            List<FlowFile> failedFlowFiles = new ArrayList<>();
+            List<FlowFile> successfulFlowFiles = new ArrayList<>();
+
+            // Prepare batch of records
+            for (int i = 0; i < flowFiles.size(); i++) {
+                FlowFile flowFile = flowFiles.get(i);
+
+                final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                session.exportTo(flowFile, baos);
+                PutRecordsRequestEntry record = new PutRecordsRequestEntry().withData(ByteBuffer.wrap(baos.toByteArray()));
+
+                String partitionKey = context.getProperty(PutKinesis.KINESIS_PARTITION_KEY)
+                    .evaluateAttributeExpressions(flowFiles.get(i)).getValue();
+
+                if ( ! StringUtils.isBlank(partitionKey) ) {
+                    record.setPartitionKey(partitionKey);
+                } else {
+                    record.setPartitionKey(Integer.toString(randomParitionKeyGenerator.nextInt()));
+                }
+
+                records.add(record);
+            }
+
+            if ( records.size() > 0 ) {
+
+                PutRecordsRequest putRecordRequest = new PutRecordsRequest();
+                putRecordRequest.setStreamName(streamName);
+                putRecordRequest.setRecords(records);
+                PutRecordsResult results = client.putRecords(putRecordRequest);
+
+                List<PutRecordsResultEntry> responseEntries = results.getRecords();
+                for (int i = 0; i < responseEntries.size(); i++ ) {
+                    PutRecordsResultEntry entry = responseEntries.get(i);
+                    FlowFile flowFile = flowFiles.get(i);
+
+                    Map<String,String> attributes = new HashMap<>();
+                    attributes.put(AWS_KINESIS_SHARD_ID, entry.getShardId());
+                    attributes.put(AWS_KINESIS_SEQUENCE_NUMBER, entry.getSequenceNumber());
+
+                    if ( ! StringUtils.isBlank(entry.getErrorCode()) ) {
+                        attributes.put(AWS_KINESIS_ERROR_CODE, entry.getErrorCode());
+                        attributes.put(AWS_KINESIS_ERROR_MESSAGE, entry.getErrorMessage());
+                        flowFile = session.putAllAttributes(flowFile, attributes);
+                        failedFlowFiles.add(flowFile);
+                    } else {
+                        flowFile = session.putAllAttributes(flowFile, attributes);
+                        successfulFlowFiles.add(flowFile);
+                    }
+                }
+                if ( failedFlowFiles.size() > 0 ) {
+                    session.transfer(failedFlowFiles, REL_FAILURE);
+                    getLogger().error("Failed to publish to kinesis {} records {}", new Object[]{streamName, failedFlowFiles});
```

Comment:
IMO a good balance is a count and an extract of 10 records (or whatever small number)